File: 2016/12/solution.js
const fs = require('fs');

const input = fs.readFileSync('input.txt', 'utf-8', (err, data) => {
  if (err) {
    console.log(err)
  } else {
    return data;
  }
}).split('\n');


const instructions = [];

for (const line of input) {
  const parts = line.split(' ');
  for (let i = 0; i < parts.length; i++) {
    const n = Number(parts[i]);
    if (!isNaN(n)) {
      parts[i] = n;
    }
  }
  instructions.push(parts);
}

const execute = (c) => {
  const register = new Map();
  register.set('a', 0);
  register.set('b', 0);
  register.set('c', c);
  register.set('d', 0);
  let i = 0;
  while (i < input.length) {
    const instruction = instructions[i];
    if (instruction[0] === 'cpy') {
      if (typeof instruction[1] === 'string') {
        register.set(instruction[2], register.get(instruction[1]));
      } else {
        register.set(instruction[2], instruction[1]);
      }
    } else if (instruction[0] === 'inc') {
      register.set(instruction[1], register.get(instruction[1]) + 1);
    } else if (instruction[0] === 'dec') {
      register.set(instruction[1], register.get(instruction[1]) - 1);
    } else {
      let check;
      if (typeof instruction[1] === 'string') {
        check = register.get(instruction[1]);
      } else {
        check = instruction[1];
      }
      if (check !== 0) {
        i += instruction[2] - 1;
      }
    }
    i++;
  }
  return register.get('a');
}

console.log(execute(0));
console.log(execute(1));

File: 2016/12/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := parseInput(strings.Split(string(data), "\n"))
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

type Instruction struct {
	instruc    string
	isAddress1 bool
	isAddress2 bool
	address1   string
	address2   string
	value1     int
	value2     int
}

func NewInstruction(s string) (*Instruction, error) {
	parts := strings.Split(s, " ")
	var isAddress1, isAddress2 bool
	var address1, address2 string
	var value1, value2 int
	if parts[0] == "cpy" || parts[0] == "jnz" {
		n1, err := strconv.Atoi(parts[1])
		if err != nil {
			isAddress1 = true
			address1 = parts[1]
		} else {
			value1 = n1
		}
		n2, err := strconv.Atoi(parts[2])
		if err != nil {
			isAddress2 = true
			address2 = parts[2]
		} else {
			value2 = n2
		}
	} else if parts[0] == "inc" || parts[0] == "dec" {
		isAddress1 = true
		address1 = parts[1]
	} else {
		return nil, fmt.Errorf("unexpected instruction: %s", s)
	}

	return &Instruction{parts[0], isAddress1, isAddress2, address1, address2, value1, value2}, nil
}

func parseInput(data []string) []*Instruction {
	puzzleInput := make([]*Instruction, len(data))
	for i, s := range data {
		instruction, err := NewInstruction(s)
		if err != nil {
			panic(err)
		}
		puzzleInput[i] = instruction
	}
	return puzzleInput
}

func execute(puzzleInput []*Instruction, cValue int) int {
	registers := make(map[string]int, 4)
	registers["c"] = cValue
	i := 0
	for i >= 0 && i < len(puzzleInput) {
		curr := puzzleInput[i]
		x := curr.value1
		if curr.isAddress1 {
			x = registers[curr.address1]
		}
		y := curr.value2
		if curr.isAddress2 {
			y = registers[curr.address2]
		}
		switch curr.instruc {
		case "cpy":
			registers[curr.address2] = x
		case "inc":
			registers[curr.address1]++
		case "dec":
			registers[curr.address1]--
		case "jnz":
			if x != 0 {
				i += y
				continue
			}
		}
		i++
	}
	return registers["a"]
}

func part1(puzzleInput []*Instruction) int {
	return execute(puzzleInput, 0)
}

func part2(puzzleInput []*Instruction) int {
	return execute(puzzleInput, 1)
}
